### ISSUE TITLE
Subprocess pid is always 0

### ISSUE THREAD

**Issue by halffullheart** [skpm/child_process#11]
The `pid` property is always 0 in my testing.
```
var subprocess = spawn('mycmd')
subprocess.pid // => 0
```

It looks like calling `processIdentifier()` before `launch()` will give you a `0`:
```
var task = NSTask.alloc().init()
task.processIdentifier() // => 0
```

`result.pid` is being set to `String(task.processIdentifier())` but this happens before `launch()` is called because that is done via `setImmediate()` here: https://github.com/skpm/child_process/blob/0513aed499b310c34ef08423275a30d9d9405912/lib/spawn.js#L216

**Comment by mathieudutour** [comment]
Ah right, thanks for the nice bug report!

A potential fix would be to use a getter instead of assigning the value directly. Want to send a PR?

**Comment by halffullheart** [comment]
Yes, I would be happy to try to put together a fix. Can you help me understand why `setImmediate()` is being used?

From what I can tell, the node API returns a subprocess with the correct PID synchronously, e.g. this works:

```
const { spawn } = require('child_process')
console.log(spawn('ls').pid) // => non-zero number
```

Using a getter to call `processIdentifier()` would still return 0 in a case like this, right?

**Comment by mathieudutour** [comment]
Hum you're right. Tbh I can't remember why I wrapped it in a `setImmediate`. 
I'm guessing it had to do with making sure the callback is called at the end of the event loop. But maybe it's not necessary

**Comment by anyab5** [comment]
Another problem with using `setImmediate` is that errors such as :
`{ Error: An Obj-C exception occurred.
	      at [native code] (/)
	      at spawn .....
	    nativeException: <NSException> Couldn't posix_spawn: error 7,
	    line: 817,
	    column: 16`
Will not get caught in the general catch flow and `error` will not get triggered so from client perspective that cannot know when and what exception occurred.
I created a PR for removing `setImmediate`

**Comment by mathieudutour** [comment]
fixed in 0.4.2
